crates/sui-types/src/transaction.rs

```diff
@@ -113,6 +122,47 @@ pub enum ObjectArg {
     Receiving(ObjectRef),
 }
 
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub enum Reservation {
+    // Reserve a specific amount of the balance.
+    MaxAmount(u64),
+    // Reserve the entire balance.
+    EntireBalance,
+}
+
+// TODO(address-balances): Rename the structs and enums.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct BalanceWithdrawArg {
+    /// The reservation of the balance to withdraw.
+    pub reservation: Reservation,
+    /// The source of the balance to withdraw.
+    pub withdraw_from: WithdrawFrom,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub enum WithdrawFrom {
+    /// Withdraw from the sender of the transaction, with balance type T as T in `Balance<T>`.
+    // TODO(address-balances): Consider hoisting the type tag to BalanceWithdrawArg.
+    Sender(TypeInput),
```

Comment:
Two things. One I still really think this `TypeInput` should be lifted to the arg itself. And two the `T` here should be `Balance<T'>` not the type argument to balance itself. This is how we will handle non-balance inputs.